polyval/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+//! POLYVAL is a GHASH-like universal hash over GF(2^128) useful for
+//! implementing AES-GCM-SIV (RFC 8452) as well as GMAC/AES-GCM.
+//!
+//! It can be thought of as the little endian equivalent of GHASH, which
+//! affords it a small performance advantage over GHASH when used on little
+//! endian architectures.
+//!
+//! By converting endianness on the way in and out, this means that POLYVAL can
+//! trivially be used to implement GHASH. This means POLYVAL can be used to
+//! implement GMAC, the Message Authentication Code used by AES-GCM.
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+
+extern crate byteorder;
+extern crate crypto_mac;
+#[cfg(feature = "zeroize")]
+extern crate zeroize;
+
+use byteorder::{BigEndian, ByteOrder, LittleEndian};
+use core::cmp::min;
+use core::marker::PhantomData;
+use crypto_mac::generic_array::{typenum::U16, GenericArray};
+use crypto_mac::{Mac, MacResult};
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// The POLYVAL universal hash function
+pub type Polyval = PolyvalCore<LittleEndian>;
+
+/// The GHASH universal hash function
+pub type GHash = PolyvalCore<BigEndian>;
```

Comment:
This seemed clever at the time, but I think I'd rather drop it and make `GHash` a newtype of `Polyval` which handles the endianness conversions. This approach is almost certainly incorrect/naive, as I believe the transformation between them involves an operation over GHASH's field.

I will delete this for now...